Add animal classes, interfaces, and polymorphism demo to ZooSimulator

Kseniya1111 opened the pull request:
Added and updated abstract classes Animal, Mammal, and Bird according to requirements.

Implemented IEatable, ISleepable, and IPlayable interfaces.

Created concrete animal classes (Lion, Elephant, Eagle, Penguin) with method overrides.

Updated ZooKeeper methods to work via interfaces.

Refactored ZooSimulator to demonstrate subtype polymorphism with Animal[] and interface-based iteration.

Kseniya1111 committed:
Interfaces has been added

rodikno commented on .DS_Store:
I was wondering what are these files :)

The .DS_Store file is a hidden system file created by macOS in folders to store custom attributes of a folder, such as the position of icons, window size, and view options. It is not related to source code or project functionality—it's purely for the Finder application on Mac computers.

In summary:
.DS_Store files are safe to delete and are not needed in a code repository. It's best practice to add .DS_Store to your .gitignore file to prevent it from being committed to your repository.

Kseniya1111 replied:
Got it, thanks!

rodikno commented on src/main/java/com/zoo/ZooSimulator.java:
```diff
@@ -1,21 +1,91 @@
+import java.util.Scanner;
+import com.zoo.animals.Animal;
+import com.zoo.animals.IPlayable;
+import com.zoo.species.Eagle;
```
Just for simplification, these imports can be simplified to `com.zoo.species.*` as you're importing all of the classes separately anyway (not a mistake just will save 3 lines)

rodikno commented on src/main/java/com/zoo/ZooSimulator.java:
```diff
@@ -1,21 +1,91 @@
+
+        System.out.println("До побачення! Дякуємо, що були з нами у зоопарку 🐾");
+
+        System.out.println(" А тепер створюю масив та ще демонструю поліморфізм підтипів ");
```
🥇

rodikno commented on src/main/java/com/zoo/ZooSimulator.java:
```diff
@@ -1,21 +1,91 @@
+            animal.makeSound();
+            animal.eat();
+            animal.sleep();
+        }
```
Ну хіба не чудово? :)

Якщо ти наслідник Animal - значить точно як мінімум матимеш ці методи - клааас 👍

rodikno commented on src/main/java/com/zoo/animals/Animal.java:
```diff
@@ -1,4 +1,56 @@
+    protected String name;
+   // protected int age;
+   // protected double weight;
+    private int energyLevel = 0;
```
Тут не дуже зрозуміло, чому дефолтним значенням ставиться саме 0 (до речі воно ж вже було нулем неявно бо поле має примітивний тип `int`) - щоб потім нижче в конструкторі поставити його в 100?

Якщо хочеться задати дефолтне значення - **має сенс задавати його прямо там де об'являємо поле** - тоді не буде шансу забути засетити його в 100 коли, наприклад, у нас будуть додаватися оверлоади для конструктору

rodikno commented on src/main/java/com/zoo/animals/Animal.java:
```diff
@@ -1,4 +1,56 @@
+        this.name = name;
+        //this.age = age;
+      //  this.weight = weight;
+        this.energyLevel = 100;
```
Вище я про ось це - якщо хочемо, щоб тваринка при народженні мала 100 енергії за замовчуванням - ідеальним способом буде саме **_set a default value to a field_**

`private int energyLevel = 100;`

rodikno commented on src/main/java/com/zoo/animals/Animal.java:
```diff
@@ -1,4 +1,56 @@
+    }
+
+
+    public abstract void sleep();
```
не помилка, але скоріше стилістична помарочка - можливо мало б сенс трошки краще все організувати в класі, зібрати всі сигнатури абстрактних методів нагорі класу, потім вже неабстрактні методи згруповані по функціоналу або ступеням приватності, приватні методи класу в самому низу

Так воно краще читається, коли все добро зібране по групах)

rodikno commented on src/main/java/com/zoo/animals/Animal.java:
```diff
@@ -1,4 +1,56 @@
+    public abstract void makeSound();
+    public abstract void eat ();
+
+     public void feedbackToPlay() {
```
Тут можливо неймінгова неточність: чи має сенс робити `feedbackToPlay()` якщо можна просто зробити `play()`?

Додатково, в тебе тепер є інтерфейс `IPlayable` в якому є абстрактні методи-контракти `play()` і `feedBackToPlay()`

Питання 1: якщо Animal має `feedbackToPlay()`, то чому вона вже тоді не імплементує IPlayable? (тут вказую на те, що подібні "автономні" вкраплення які дублюють логіку задану інтерфейсом можуть надалі приносити проблеми - краще уніфікувати і законтрактувати через імплементацію інтерфейсу (вважай, якщо в рамках нормальності для тварини гратися - тоді нехай вона в тому числі в усьому іншому поводить себе як така, що може гратися, що нашою мовою значить `implements IPlayable`)

Питання 2: але воно більше до самого сенсу існування та обраного імені цього методу - чому довелося робити `play()` і `feedbackToPlay()` окремими методами? Чи є логічний зв*язок між ними?
На перший погляд, інтуїтивно, `play()` це публічна поведінка тварини, а `feedbackToPlay()` це завершаюча частина поведінки `play()`, відповідно `feedbackToPlay()` для мене **відчувається** як метод, який повинен бути приватним

Наприклад машина може публічно `drive()`, а всередині `drive()` викликаються приватні більш специфічні `pumpFuelIntoIgnitionChamber()`, `ignite()`, `turnCrankShaft()`, `spinWheels()` і т.п.


Якщо будуть питання по цьому - нагадай на занятті - розглянемо детальніше :)

rodikno commented on src/main/java/com/zoo/animals/Animal.java:
```diff
@@ -1,4 +1,56 @@
-}
+public  abstract class Animal implements IEatable,ISleepable {
+    protected String name;
+   // protected int age;
```
А чому вирішила закоментувати ці два поля?
Наслідники такі як `Bird` і `Mammal` всеодно мають мати вік і вагу, просто тепер кожен з них буде вимушений декларувати ці поля сам - а могли б просто наслідувати протектед поля від `Animal`

Та і це не порушувало б логіки - 

**Animal has age and weight**
Bird **is** an Animal, hence it has age and weight
Mammal **is** an Animal, hence it has age and weight

rodikno commented on src/main/java/com/zoo/animals/Bird.java:
```diff
@@ -1,4 +1,26 @@
+    }
+
+    public void fly()
+    { changeEnergy(-15);
```
Formatting! ^_^

`Ctrl + Alt + Shift + L`

rodikno commented on src/main/java/com/zoo/animals/IEatable.java:
```diff
@@ -0,0 +1,7 @@
+     void eat ();
+     void changeEnergy(int amount);
+     String getName();
+}
```
Здається, в завданні такого не було :)

В чому помилка: інтерфейси задають абстрактний контракт: **який я? що я роблю? що я можу?**

Якщо я імплементую контракт під назвою "Я ТОЙ, ЩО МОЖЕ ЇСТИ", то чому цей контракт мене зобов*язує ще вміти змінювати рівень енергії і повертати власне ім'я?

Інтерфейси повинні містити необхідний мінімум методів, які даватимуть їх імплементаторам називатися "Я ТОЙ, ЩО ВМІЄ ЇСТИ", "Я ТОЙ, ЩО ВМІЄ ГРАТИСЯ", "Я ТОЙ, ЩО ВМІЄ СПАТИ"

Підказка: якщо я кажу, що вмію їсти, то я не підписуюсь цим твердженням на те, що я вмію вимовляти власне ім'я і повертати рівень енергії: так, я можу робити і це і багато ще чого, але підписуючись під "вмію їсти" я даю обов'язок вміти безпосередньо їсти.

rodikno commented on src/main/java/com/zoo/animals/IPlayable.java:
```diff
@@ -0,0 +1,8 @@
+public interface IPlayable {
+    void play ();
+    void changeEnergy(int amount);
+    String getName();
```
Як тепер видно, додавання в інтерфейс "сторонніх фіч" типу `changeEnergy()`, `getName()` - приводить як мінімум до дублювання, і це ще не найгірше, що могло трапитись

Підказка: залишити тільки `play()`, все інше нехай визначають класи-імплементатори (якщо їм треба), або інші інтерфейси які будуть мати відповідне призначення

rodikno commented on src/main/java/com/zoo/animals/ISleepable.java:
```diff
@@ -0,0 +1,5 @@
+package com.zoo.animals;
+
+public interface ISleepable {
+     void sleep  ();
```
Formatting ^_^

`void sleep();` not `void sleep   ()`

`Ctrl + Alt + Shift + L`

rodikno commented on src/main/java/com/zoo/animals/Mammal.java:
```diff
@@ -1,4 +1,36 @@
-}
+public abstract class Mammal extends Animal {
+    boolean isGroomLong;
+    int protect;
```
Не дуже вдала назва змінної-поля: вже далі по коду я бачу, що тут зберігатиметься ступінь захисту (який вже також далі видно залежить від довжини шерсті)

Тоді має сенс назвати це поле `selfProtectionLevel` чи приблизно так, щоб при першому ж погляді на нього було зрозуміло, навіщо воно потрібне - це було б прям топчик :)

rodikno commented on src/main/java/com/zoo/animals/Mammal.java:
```diff
@@ -1,4 +1,36 @@
+    public void groom() {
+        if (isGroomLong) {
+            protect = protect * 2;
+        } else
```
Formatting ^_^, якщо else без фігурних - тоді вже в один рядок

Але тут рекомендував би для консистентності - якщо `if` з тілом в фігурних скобках - тоді і його `else` теж нехай матиме тіло з фігурних скобок - так воно набагато легше сприймається при читанні

rodikno commented on src/main/java/com/zoo/species/Eagle.java:
```diff
@@ -1,4 +1,51 @@
+import com.zoo.animals.IPlayable;
+import com.zoo.animals.ISleepable;
+
+public   class Eagle extends Bird  implements IEatable, IPlayable, ISleepable {
```
Орел це Птах, Птах це Тварина, Тварина вже імплементує IPlayable, ISleepable

відповідно, навіщо тоді нам явно тепер казати, що птах `implements IPlayable, ISleepable?`

по факту це як начебто "порушення субординації" - птах всеодно буде змушений мати властиві інтерфейсам методи якщо хоча б хтось з його предків імплементує ці інтерфейси.

Якщо усі предки вище абстрактні - значить Птах першим дійсно розкаже, як саме він грається і спить
Якщо вище у Птаха є хоча б один неабстрактний предок - значить він вже **змушений** був вказати як саме він грається і спить - і тепер у Птаха є Вибір - робити так як батьківський клас чи надати власні оверрайди

rodikno commented on src/main/java/com/zoo/species/Eagle.java:
```diff
@@ -1,4 +1,51 @@
+    public void play ()
+    {
+        changeEnergy(-20);
+        System.out.println("Іііііііііі! - каже лев коли погрався");
```
тут одруківочка - це ж у нас птах, а не лев :)

rodikno commented on src/main/java/com/zoo/species/Elephant.java:
```diff
@@ -1,4 +1,59 @@
+    public void move ()
+    {
+        changeEnergy(-40);
+        System.out.println("Слон вирішив прогулятися та підтримати соціальні зв'язки");
```
Це чудово, що у нас такий соціалізований слон :)

Але дивлячись загалом - яка різниця між поведінкою `move()` і `walk()`? Якщо слон рухається, значить він ходить? Тоді чому у нас два різних методи для одного і того ж? (причому вони ще на енергію впливають по-різному, пройтися коштує 70 енергії, а "поворухнутися" -40 енергії

rodikno commented on src/main/java/com/zoo/species/Lion.java:
```diff
@@ -1,4 +1,53 @@
+import com.zoo.animals.ISleepable;
+import com.zoo.animals.Mammal;
+
+public class Lion  extends Mammal  implements IEatable, IPlayable, ISleepable {
```
Тут також - якщо хтось з батьків вже імплементує якісь інтерфейси - немає потреби явно вказувати, що клас-дитинча їх імплементує - усі потрібні методи і так прийдуть від когось з батьків по ієрархії наслідування

rodikno commented on src/main/java/com/zoo/species/Lion.java:
```diff
@@ -1,4 +1,53 @@
+
+public class Lion  extends Mammal  implements IEatable, IPlayable, ISleepable {
+
+    public Lion(String name, int age, double weight,boolean isGroomLong,int protect,String furColor) {
```
Formatting ^_^

Як казав один полковник - краще навіть безобразно, але однообразно

В нашому випадку десь пробіл після коми є, десь - немає, бажано щоб пробіл після коми був аж до кінця списку параметрів

rodikno commented on src/main/java/com/zoo/species/Penguin.java:
```diff
@@ -1,4 +1,60 @@
+    private void swim ()
+    {
+        changeEnergy(-20);
+        System.out.println("Пінгвін поплавав  за рибами у льодяній воді.");
```
Частіше має сенс приватні методи визначати в саамому низу класу після усіх публічних як мінімум - тоді важче заплутатись

rodikno commented on src/main/java/com/zoo/zookeper/ZooKeeper.java:
```diff
@@ -1,4 +1,47 @@
+
+
+    public void playWithAnimal(IPlayable animal)
+    { animal.changeEnergy(-25);
```
ми ж тут очікуємо `IPlayable` параметр - так змусимо його просто гратися

`animal.play()` - це все, що ми можемо від нього вимагати
все інше - ЯК САМЕ він грається - це вже його внутрішні справи - ми як зовнішній агент не маємо вказувати тваринці як гратися, скільки енергії їй витрачати і як давати фідбек

(це буде перекликатися з виправленнями вище - тобі треба буде поправити усю ієрархію, переробити самі визначення інтерфейсів) - і тоді тут не було б можливості допустити цю помилку, бо тобі доступен був би тільки метод `play()`

rodikno commented on src/main/java/com/zoo/zookeper/ZooKeeper.java:
```diff
@@ -1,4 +1,47 @@
+
+    }
+
+    public  ZooKeeper (String name) {
```
коснтруктори зазвичай визначаємо нагорі класу, щоб їх було одразу видно коли хтось читає код

rodikno commented on src/main/java/com/zoo/zookeper/ZooKeeper.java:
```diff
@@ -1,4 +1,47 @@
+    public  ZooKeeper (String name) {
+        this.name = name;
+    }
+    public String getZooKeeperName() {
```
Неймінг трохи надлишковий

Цей метод знаходиться в класі `ZooKeeper`, і коли ми будемо викликати його від імені об'єкту-наглядача ми вже знатимемо, що це саме об'єкт ZooKeeper'а

Тому тут прям просто і зрозуміло: поле називається `name` - значить метод-геттер для нього `getName()` - це конвенція

Це дозволяє уникнути таких своєрідних "тавтологій", щоб було зрозуміло різницю:

```java
Наглядач мійНаглядач = new Наглядач("Андрій");

мійНаглядач.отриматиІмяНаглядача(); //тут тавтологізм, бо зрозуміло що це наглядач же

//краще ось так
мійНаглядач.отриматиІмя()
```

rodikno reviewed (changes requested):
Треба трошки "допилити", відписав коментами

Після внесення змін - запуш в ту ж гілку пліз і напиши приватним коментарем в домашці №8 і №9, що зміни внесено + поміть домашку виконаною знову

Kseniya1111 committed:
changes which added after teacher review

Kseniya1111 committed:
ZooSimulator has been updated

Kseniya1111 committed:
Changes which has been added after teacher's review

-Interfaces have been changed and simplified
-Formatting has been added
-Programming logic has been changed

rodikno commented on src/main/java/com/zoo/animals/Animal.java:
```diff
@@ -1,4 +1,39 @@
-}
+public abstract class Animal implements IEatable, ISleepable, IPlayable {
+    protected String name;
+    private int energyLevel = 100;
```
nice!

rodikno commented on src/main/java/com/zoo/animals/Animal.java:
```diff
@@ -1,4 +1,39 @@
+
+    public void changeEnergy(int amount) {
+        energyLevel += amount;
+        if (energyLevel > 100) {
```
One additional validation can be done, ensuring the energy also can't be less than 0

rodikno replied:
else if block I mean

rodikno commented on src/main/java/com/zoo/animals/Animal.java:
```diff
@@ -1,4 +1,39 @@
+    }
+
+    public void displayInfo() {
+        System.out.println("Info about current  animal: Name: " + name);
```
For a better UX -> more fields can be added: age, weight etc

rodikno commented on src/main/java/com/zoo/animals/Mammal.java:
```diff
@@ -1,4 +1,24 @@
+
+    public Mammal(String name, boolean isGroomLong, int protect, String furColor, int age, double weight) {
+        super(name);
+        this.hasSpecialProtection = isGroomLong;
```
This is a bit misleading -> I'd say parameter name should match the corresponding property name

rodikno commented on src/main/java/com/zoo/animals/Mammal.java:
```diff
@@ -1,4 +1,24 @@
+    public Mammal(String name, boolean isGroomLong, int protect, String furColor, int age, double weight) {
+        super(name);
+        this.hasSpecialProtection = isGroomLong;
+        this.protect = protect;
```
Name is not fully clear -> protect sounds like a verb, but it's an adjective here

so `protectionLevel` or smth like this can provide more clarity

rodikno commented on src/main/java/com/zoo/species/Elephant.java:
```diff
@@ -1,4 +1,57 @@
+    @Override
+    public void groom() {
+        protect += 30;
+        System.out.println("Слон потоптав землю і підняв пил — тепер його важко дістати. Захист: " + protect);
```
<3

rodikno reviewed (approved):
Good job! Approved :)
